### PR TITLE
BalancerWeightedPoolSwapper

### PR DESCRIPTION
**Pull request by bogdoslavik**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -57,11 +59,11 @@
     "hardhat": "^2.8.3",
     "hardhat-abi-exporter": "^2.8.0",
     "hardhat-contract-sizer": "^2.1.1",
+    "hardhat-deploy": "^0.9.28",
```
Comment (belbix) [inline package.json:62]: I think we don;t need this package, can remove

File: scripts/utils/DeployerUtils.ts [hunk 2]
```diff
@@ -146,4 +155,72 @@ export class DeployerUtils {
     }
   }
 
+  public static pauseWindowDuration = BigNumber.from(90 * 24 * 3600);
```
Comment (belbix) [inline scripts/utils/DeployerUtils.ts:158]: move to deployBalancer() attributes with default values

File: contracts/swappers/BalancerSwapper.sol [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+import "../interfaces/IBVault.sol";
+import "../interfaces/IBWeightedPoolMinimal.sol";
+import "../proxy/ControllableV3.sol";
+import "../openzeppelin/Math.sol";
+import "../lib/WeightedMath.sol";
+
+/// @title Swap tokens via Balancer vault.
+/// @notice !!! Weighted pools only supported yet !!!
+/// @author bogdoslav
+contract BalancerSwapper is ControllableV3, ISwapper {
```
Comment (belbix) [inline contracts/swappers/BalancerSwapper.sol:18]: rename to BalancerWeightedPoolSwapper

File: contracts/swappers/BalancerSwapper.sol [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+import "../interfaces/IERC20Metadata.sol";
+import "../interfaces/ISwapper.sol";
+import "../interfaces/IBVault.sol";
+import "../interfaces/IBWeightedPoolMinimal.sol";
+import "../proxy/ControllableV3.sol";
+import "../openzeppelin/Math.sol";
+import "../lib/WeightedMath.sol";
+
+/// @title Swap tokens via Balancer vault.
+/// @notice !!! Weighted pools only supported yet !!!
```
Comment (belbix) [inline contracts/swappers/BalancerSwapper.sol:16]: I guess we will have dedicated swapper for each pool type

File: contracts/swappers/BalancerSwapper.sol [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+  ) public view override returns (uint) {
+    { // take pool commission
+      uint swapFeePercentage = IBWeightedPoolMinimal(pool).getSwapFeePercentage();
+      amount -= amount * swapFeePercentage / 10**18;
+    }
+    bytes32 poolId = IBWeightedPoolMinimal(pool).getPoolId();
+    (IERC20[] memory tokens,
+    uint256[] memory balances,) = IBVault(balancerVault).getPoolTokens(poolId);
+
+    require(tokens.length == 2, 'Wrong pool tokens length');
```
Comment (belbix) [inline contracts/swappers/BalancerSwapper.sol:87]: why we can not use it for pools with multiple tokens?

File: contracts/swappers/BalancerSwapper.sol [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    require(
+      (tokens[0] == IERC20(tokenIn) && tokens[1] == IERC20(tokenOut)) ||
```
Comment (belbix) [inline contracts/swappers/BalancerSwapper.sol:89]: I guess you can check tokens in a loop on existence

File: contracts/swappers/BalancerSwapper.sol [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+    bool direct = tokens[0] == IERC20(tokenIn);
+
+    (uint weightIn, uint weightOut) = direct ?
+      (weights[0], weights[1]) : (weights[1], weights[0]);
+
+    (uint balanceIn, uint balanceOut) = direct ?
+      (balances[0], balances[1]) : (balances[1], balances[0]);
+
+    return WeightedMath._calcOutGivenIn(
```
Comment (belbix) [inline contracts/swappers/BalancerSwapper.sol:105]: looks like this function will support a pool with multiple tokens

File: contracts/swappers/BalancerSwapper.sol [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+      amount
+    );
+  }
+
+  // *************************************************************
+  //                        SWAP
+  // *************************************************************
+
+  /// @dev Swap given tokenIn for tokenOut. Assume that tokenIn already sent to this contract.
+  /// @param pool Dystopia pool
```
Comment (belbix) [inline contracts/swappers/BalancerSwapper.sol:119]: fix comment

**Review by belbix — CHANGES_REQUESTED** [review]

File: contracts/swappers/BalancerSwapper.sol (hunk 5 above)
Comment (bogdoslavik) [reply contracts/swappers/BalancerSwapper.sol]: It is compatible. Unnecessary check removed. Code improved.

**Pull request merged**